Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+        }
+    }
+    private var lookupWebViewHandlersByLookupSettings: [LookupWebViewHandler.LookupSettings: LookupWebViewHandler] = [:]
+    
+    // MARK: Object Lifecycle
+    init(
+        dbStorage: DbStorage,
+        fileStorage: FileStorage,
+        translatorsController: TranslatorsAndStylesController,
+        schemaController: SchemaController,
+        dateParser: DateParser,
+        remoteFileDownloader: RemoteAttachmentDownloader
+    ) {
+        self.fileStorage = fileStorage
+        self.dbStorage = dbStorage
+        self.translatorsController = translatorsController
+        self.schemaController = schemaController
+        self.dateParser = dateParser
+        self.remoteFileDownloader = remoteFileDownloader
+        
+        self.accessQueue = DispatchQueue(label: "org.zotero.IdentifierLookupController.accessQueue", qos: .userInteractive, attributes: .concurrent)
+        self.backgroundQueue = DispatchQueue(label: "org.zotero.IdentifierLookupController.backgroundProcessing", qos: .userInitiated)
+        self.observable = PublishSubject()
+        self.disposeBag = DisposeBag()
+        
+        setupObservers()
+    }
+    
+    // MARK: Actions
+    func initialize(libraryId: LibraryIdentifier, collectionKeys: Set<String>, completion: @escaping ([LookupData]?) -> Void) {
+        accessQueue.async(flags: .barrier) { [weak self] in
+            var lookupData: [LookupData]?
+            defer {
+                completion(lookupData)
+            }
+            guard let self = self else { return }
+            let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+            if self.lookupWebViewHandlersByLookupSettings[lookupSettings] != nil {
+                lookupData = self.lookupData
+                return
+            }
+            inMainThread(sync: true) {
+                if let webView = self.webViewProvider?.addWebView() {
+                    let lookupWebViewHandler = LookupWebViewHandler(lookupSettings: lookupSettings, webView: webView, translatorsController: self.translatorsController)
+                    self.lookupWebViewHandlersByLookupSettings[lookupSettings] = lookupWebViewHandler
+                }
+            }
+            guard let lookupWebViewHandler = self.lookupWebViewHandlersByLookupSettings[lookupSettings] else {
+                DDLogError("IdentifierLookupController: can't create LookupWebViewHandler instance")
+                return
+            }
+            self.setupObserver(for: lookupWebViewHandler)
+            lookupData = self.lookupData
+        }
+    }
+    
+    func lookUp(libraryId: LibraryIdentifier, collectionKeys: Set<String>, identifier: String) {
+        let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+        lookupWebViewHandlersByLookupSettings[lookupSettings]?.lookUp(identifier: identifier)
```

Comment:
Does this always exist? Maybe add a `guard` and a `DDLogError` at least in case something breaks in future and users report that this does nothing.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/97438bb929d4e0d770d16970d92326c45fac536b